A serverConnection that has already been set to disconnected is still processing requests

### Search before asking

- [x] I searched in the [issues](https://github.com/apache/fluss/issues) and found nothing similar.


### Fluss version

main (development)

### Please describe the bug 🐞

A `ServerConnection` that has already been set to `DISCONNECTED` is still processing requests.  

The error occurs as follows:
```java
org.apache.fluss.server.replica.fetcher.ReplicaFetcherThread [] - Error in response for fetch log request org.apache.fluss.rpc.messages.FetchLogRequest@691b6a8b
java.util.concurrent.ExecutionException: org.apache.fluss.exception.NetworkException: org.apache.fluss.exception.DisconnectException: Cannot send request to server xxx (id: ts-131, rack: xxx) because it is disconnected.
	at java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:396) ~[?:?]
	at java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2096) ~[?:?]
	at org.apache.fluss.server.replica.fetcher.ReplicaFetcherThread.processFetchLogRequest(ReplicaFetcherThread.java:227) ~[fluss-server-0.8-SNAPSHOT.jar:0.8-SNAPSHOT]
	at java.util.Optional.ifPresent(Optional.java:178) [?:?]
	at org.apache.fluss.server.replica.fetcher.ReplicaFetcherThread.maybeFetch(ReplicaFetcherThread.java:153) [fluss-server-0.8-SNAPSHOT.jar:0.8-SNAPSHOT]
	at org.apache.fluss.server.replica.fetcher.ReplicaFetcherThread.doWork(ReplicaFetcherThread.java:124) [fluss-server-0.8-SNAPSHOT.jar:0.8-SNAPSHOT]
	at org.apache.fluss.utils.concurrent.ShutdownableThread.run(ShutdownableThread.java:96) [fluss-server-0.8-SNAPSHOT.jar:0.8-SNAPSHOT]
Caused by: org.apache.fluss.exception.NetworkException: org.apache.fluss.exception.DisconnectException: Cannot send request to server xxx (id: ts-131, rack: xxx) because it is disconnected.
Caused by: org.apache.fluss.exception.DisconnectException: Cannot send request to server xxx (id: ts-131, rack: xxx) because it is disconnected.
```
This error is unexpected. In theory, once a `serverConnection` is set to `DISCONNECTED`, it must have reached `serverConnection.close()`, which should definitely trigger either `closeFuture.complete()` or `closeFuture.completeExceptionally()`. However, the current behavior suggests that neither of these methods is being called. At present, we strongly suspect that the issue lies in `channel.close().addListener()` not being invoked as expected.

### Solution

_No response_

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR!

--- a/fluss-rpc/src/main/java/org/apache/fluss/rpc/netty/client/ServerConnection.java
+++ b/fluss-rpc/src/main/java/org/apache/fluss/rpc/netty/client/ServerConnection.java
@@ -17,6 +17,7 @@
 
 package org.apache.fluss.rpc.netty.client;
 
+import org.apache.fluss.annotation.VisibleForTesting;
 import org.apache.fluss.cluster.ServerNode;
 import org.apache.fluss.exception.DisconnectException;
 import org.apache.fluss.exception.FlussRuntimeException;
@@ -170,39 +171,18 @@ final class ServerConnection {
             }
 
             if (channel != null) {
-                channel.close()
-                        .addListener(
-                                (ChannelFutureListener)
-                                        future -> {
+                // Close the channel directly, without waiting for the channel to close properly.
+                channel.close();
+            }
 
-                                            // when finishing, if netty successfully closes the
-                                            // channel, then the provided exception is used as
-                                            // the reason for the closing. If there was something
-                                            // wrong at the netty side, then that exception is
-                                            // prioritized over the provided one.
-                                            if (future.isSuccess()) {
-                                                if (cause instanceof ClosedChannelException) {
-                                                    // the ClosedChannelException is expected
-                                                    closeFuture.complete(null);
-                                                } else {
-                                                    closeFuture.completeExceptionally(cause);
-                                                }
-                                            } else {
-                                                LOG.warn(
-                                                        "Something went wrong when trying to close connection due to : ",
-                                                        cause);
-                                                closeFuture.completeExceptionally(future.cause());
-                                            }
-                                        });
+            // TODO all return completeExceptionally will let some test cases blocked, so we
+            // need to find why the test cases are blocked and remove the if statement.
+            if (cause instanceof ClosedChannelException
+                    || cause.getCause() instanceof ConnectException) {
+                // the ClosedChannelException and ConnectException is expected.
+                closeFuture.complete(null);
             } else {
-                // TODO all return completeExceptionally will let some test cases blocked, so we
-                // need to find why the test cases are blocked and remove the if statement.
-                if (cause.getCause() instanceof ConnectException) {
-                    // the ConnectException is expected
-                    closeFuture.complete(null);
-                } else {
-                    closeFuture.completeExceptionally(cause);
-                }
+                closeFuture.completeExceptionally(cause);
             }
 
             connectionMetricGroup.close();
@@ -491,7 +471,8 @@ final class ServerConnection {
      * <li>READY: connection is ready to send requests.
      * <li>DISCONNECTED: connection is failed to establish.
      */
-    private enum ConnectionState {
+    @VisibleForTesting
+    enum ConnectionState {
         CONNECTING,
         CHECKING_API_VERSIONS,
         AUTHENTICATING,
@@ -564,5 +545,10 @@ final class ServerConnection {
         public String ipAddress() {
             return ((InetSocketAddress) channel.remoteAddress()).getAddress().getHostAddress();
         }
+    }
+
+    @VisibleForTesting
+    ConnectionState getConnectionState() {
+        return state;
     }
 }

--- a/fluss-rpc/src/test/java/org/apache/fluss/rpc/netty/client/NettyClientTest.java
+++ b/fluss-rpc/src/test/java/org/apache/fluss/rpc/netty/client/NettyClientTest.java
@@ -42,6 +42,7 @@ import org.junit.jupiter.api.AfterEach;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
 
+import java.net.ConnectException;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Collections;
@@ -156,9 +157,9 @@ final class NettyClientTest {
                                 nettyClient
                                         .sendRequest(serverNode, ApiKeys.API_VERSIONS, request)
                                         .get())
-                .isInstanceOf(ExecutionException.class)
-                .hasMessageContaining("Disconnected from node")
-                .hasRootCauseMessage("finishConnect(..) failed: Connection refused");
+                .rootCause()
+                .isInstanceOf(ConnectException.class)
+                .hasMessageContaining("Connection refused");
         assertThat(nettyClient.connections().size()).isEqualTo(0);
 
         // restart the netty server.

--- a/fluss-rpc/src/test/java/org/apache/fluss/rpc/netty/client/ServerConnectionTest.java
+++ b/fluss-rpc/src/test/java/org/apache/fluss/rpc/netty/client/ServerConnectionTest.java
@@ -1,0 +1,141 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.fluss.rpc.netty.client;
+
+import org.apache.fluss.cluster.Endpoint;
+import org.apache.fluss.cluster.ServerNode;
+import org.apache.fluss.cluster.ServerType;
+import org.apache.fluss.config.Configuration;
+import org.apache.fluss.exception.DisconnectException;
+import org.apache.fluss.metrics.groups.MetricGroup;
+import org.apache.fluss.metrics.util.NOPMetricsGroup;
+import org.apache.fluss.rpc.TestingGatewayService;
+import org.apache.fluss.rpc.messages.GetTableSchemaRequest;
+import org.apache.fluss.rpc.messages.PbTablePath;
+import org.apache.fluss.rpc.metrics.TestingClientMetricGroup;
+import org.apache.fluss.rpc.netty.client.ServerConnection.ConnectionState;
+import org.apache.fluss.rpc.netty.server.NettyServer;
+import org.apache.fluss.rpc.netty.server.RequestsMetrics;
+import org.apache.fluss.rpc.protocol.ApiKeys;
+import org.apache.fluss.security.auth.AuthenticationFactory;
+import org.apache.fluss.security.auth.ClientAuthenticator;
+import org.apache.fluss.shaded.netty4.io.netty.bootstrap.Bootstrap;
+import org.apache.fluss.shaded.netty4.io.netty.channel.EventLoopGroup;
+import org.apache.fluss.utils.NetUtils;
+
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+import java.util.Collections;
+import java.util.concurrent.CompletableFuture;
+
+import static org.apache.fluss.rpc.netty.NettyUtils.getClientSocketChannelClass;
+import static org.apache.fluss.rpc.netty.NettyUtils.newEventLoopGroup;
+import static org.apache.fluss.utils.NetUtils.getAvailablePort;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+/** Test for {@link ServerConnection}. */
+public class ServerConnectionTest {
+
+    private EventLoopGroup eventLoopGroup;
+    private Bootstrap bootstrap;
+    private ClientAuthenticator clientAuthenticator;
+    private Configuration conf;
+    private NettyServer nettyServer;
+    private ServerNode serverNode;
+    private TestingGatewayService service;
+
+    @BeforeEach
+    void setUp() throws Exception {
+        conf = new Configuration();
+        buildNettyServer(0);
+
+        eventLoopGroup = newEventLoopGroup(1, "fluss-netty-client-test");
+        bootstrap =
+                new Bootstrap()
+                        .group(eventLoopGroup)
+                        .channel(getClientSocketChannelClass(eventLoopGroup))
+                        .handler(new ClientChannelInitializer(5000));
+        clientAuthenticator =
+                AuthenticationFactory.loadClientAuthenticatorSupplier(new Configuration()).get();
+    }
+
+    @AfterEach
+    void tearDown() throws Exception {
+        if (nettyServer != null) {
+            nettyServer.close();
+        }
+
+        if (eventLoopGroup != null) {
+            eventLoopGroup.shutdownGracefully();
+        }
+    }
+
+    @Test
+    void testConnectionClose() {
+        ServerConnection connection =
+                new ServerConnection(
+                        bootstrap,
+                        serverNode,
+                        TestingClientMetricGroup.newInstance(),
+                        clientAuthenticator,
+                        false);
+        ConnectionState connectionState = connection.getConnectionState();
+        assertThat(connectionState).isEqualTo(ConnectionState.CONNECTING);
+
+        GetTableSchemaRequest request =
+                new GetTableSchemaRequest()
+                        .setTablePath(
+                                new PbTablePath().setDatabaseName("test").setTableName("test"))
+                        .setSchemaId(0);
+        connection.send(ApiKeys.GET_TABLE_SCHEMA, request);
+
+        CompletableFuture<Void> future = connection.close();
+        connectionState = connection.getConnectionState();
+        assertThat(connectionState).isEqualTo(ConnectionState.DISCONNECTED);
+        assertThat(future.isDone()).isTrue();
+
+        assertThatThrownBy(() -> connection.send(ApiKeys.GET_TABLE_SCHEMA, request).get())
+                .rootCause()
+                .isInstanceOf(DisconnectException.class)
+                .hasMessageContaining("Cannot send request to server");
+        future = connection.close();
+        assertThat(future.isDone()).isTrue();
+    }
+
+    private void buildNettyServer(int serverId) throws Exception {
+        try (NetUtils.Port availablePort = getAvailablePort()) {
+            serverNode =
+                    new ServerNode(
+                            serverId, "localhost", availablePort.getPort(), ServerType.COORDINATOR);
+            service = new TestingGatewayService();
+            MetricGroup metricGroup = NOPMetricsGroup.newInstance();
+            nettyServer =
+                    new NettyServer(
+                            conf,
+                            Collections.singleton(
+                                    new Endpoint(serverNode.host(), serverNode.port(), "INTERNAL")),
+                            service,
+                            metricGroup,
+                            RequestsMetrics.createCoordinatorServerRequestMetrics(metricGroup));
+            nettyServer.start();
+        }
+    }
+}

